### PR TITLE
allow global_step to be passed as argument

### PR DESCRIPTION
**Pull request by kashif**
now all optimizers can use a decaying learning rate

for issue #635

### CODE REVIEW

**Comment by kashif** [comment]
Dear @dustinvtran kindly have a look at this PR, and do squash it if you merge eventually.

So the current change means that in all the examples the default optimizer is Adam without the decaying learning rate. So to get the old behaviour I will go through the sample code and add a  `global_step` variable and pass it along.

What do you say?

File: edward/inferences/variational_inference.py [hunk 1]
```diff
@@ -87,38 +89,38 @@ def initialize(self, optimizer=None, var_list=None, use_prettytensor=False,
 
       self.summarize = tf.summary.merge_all(key=summary_key)
 
+    if isinstance(global_step, tf.Variable):
+        starter_learning_rate = 0.1
```
Comment (dustinvtran) [inline edward/inferences/variational_inference.py]: We use two-space indent

File: edward/inferences/variational_inference.py [hunk 2]
```diff
@@ -87,38 +89,38 @@ def initialize(self, optimizer=None, var_list=None, use_prettytensor=False,
...
+        optimizer = tf.train.FtrlOptimizer(learning_rate)
       elif optimizer == 'rmsprop':
-        optimizer = tf.train.RMSPropOptimizer(0.01)
+        optimizer = tf.train.RMSPropOptimizer(learning_rate)
       else:
         raise ValueError('Optimizer class not found:', optimizer)
-
-      global_step = None
     elif isinstance(optimizer, tf.train.Optimizer):
       # Custom optimizers have no control over global_step.
```
Comment (dustinvtran) [inline edward/inferences/variational_inference.py]: This part in the logic can be removed.

**Comment by kashif** [comment]
dear @dustinvtran happy for another review

File: edward/inferences/variational_inference.py [hunk 3]
```diff
@@ -87,38 +89,36 @@ def initialize(self, optimizer=None, var_list=None, use_prettytensor=False,
 
       self.summarize = tf.summary.merge_all(key=summary_key)
 
-    if optimizer is None:
+    learning_rate = 0.01
```
Comment (dustinvtran) [inline edward/inferences/variational_inference.py]: This line should be moved to L103 after checking `isinstance(optimizer, str)`.

File: edward/inferences/variational_inference.py [hunk 4]
```diff
@@ -87,38 +89,36 @@ def initialize(self, optimizer=None, var_list=None, use_prettytensor=False,
...
+    if optimizer is None or isinstance(global_step, tf.Variable):
       # Use ADAM with a decaying scale factor.
-      global_step = tf.Variable(0, trainable=False, name="global_step")
+      if not global_step:
```
Comment (dustinvtran) [inline edward/inferences/variational_inference.py]: This logic could be flatter/less nested.

I'm also not sure if it works. For example, suppose `isinstance(global_step, tf.Variable)` is True and the user passes in `optimizer` as a `str`. Then the logic will create the decaying learning rate and exit the if-else clause without building any optimizer.

Comment (dustinvtran) [inline edward/inferences/variational_inference.py]: Suggestion: collapse the logic to be
```python
    if optimizer is None and global_step is None:
      # Default optimizer always uses a global step variable.
      global_step = tf.Variable(0, trainable=False, name="global_step")
    
    # Build learning rate step-size.
    if isinstance(global_step, tf.Variable):
      starter_learning_rate = 0.1
      learning_rate = tf.train.exponential_decay(starter_learning_rate,
                                                 global_step,
                                                 100, 0.9, staircase=True)
    else:
      learning_rate = 0.01

    # Build optimizer.
    if optimizer is None:
      optimizer = tf.train.AdamOptimizer(learning_rate)
    elif isinstance(optimizer, str):
       ...
```

**Comment by kashif** [comment]
cool will do! My computer is with tech support... I hope I can pick it up
today and then I can push it.

File: edward/inferences/variational_inference.py (hunk 4 above)
Comment (kashif) [inline edward/inferences/variational_inference.py]: cool 💻  back from the 💀 ... anyways, so 1 question, do I remove the option that the  `optimizer` is of type `tf.train.Optimizer`, e.g. from the `TypeError` string and function doc?

Since it can now be either `None` or a string.

Thanks!

**Comment by dustinvtran** [comment]
> do I remove the option that the optimizer is of type tf.train.Optimizer, e.g. from the TypeError string and function doc?
> 
> Since it can now be either None or a string.


That's correct. It should be converted back to `raise TypeError("Optimizer must be str or tf.train.Optimizer.")`

File: edward/inferences/variational_inference.py [hunk 5]
```diff
@@ -50,6 +50,8 @@ def initialize(self, optimizer=None, var_list=None, use_prettytensor=False,
       ``True`` if aim to use PrettyTensor optimizer (when using
       PrettyTensor) or ``False`` if aim to use TensorFlow optimizer.
       Defaults to TensorFlow.
+    global_step : a tf.Variable, optional
```
Comment (dustinvtran) [inline edward/inferences/variational_inference.py]: `global_step : tf.Variable, optional`

File: edward/inferences/variational_inference.py [hunk 6]
```diff
@@ -87,38 +89,41 @@ def initialize(self, optimizer=None, var_list=None, use_prettytensor=False,
...
       self.summarize = tf.summary.merge_all(key=summary_key)
 
-    if optimizer is None:
-      # Use ADAM with a decaying scale factor.
+    if optimizer is None and global_step is None:
+      # Default optimizer always uses a global step variable.
+      global_step = tf.Variable(0, trainable=False, name="global_step")
+
+    if isinstance(global_step, tf.Variable):
       global_step = tf.Variable(0, trainable=False, name="global_step")
```
Comment (dustinvtran) [inline edward/inferences/variational_inference.py]: Caught another typo. Looks like this line should be removed

File: edward/inferences/variational_inference.py [hunk 7]
```diff
@@ -87,38 +89,41 @@ def initialize(self, optimizer=None, var_list=None, use_prettytensor=False,
...
       starter_learning_rate = 0.1
       learning_rate = tf.train.exponential_decay(starter_learning_rate,
                                                  global_step,
                                                  100, 0.9, staircase=True)
+    else:
+      learning_rate = 0.01
```
Comment (dustinvtran) [inline edward/inferences/variational_inference.py]: After this line, we still have to set `global_step = None`.

**Comment by kashif** [comment]
thank you @dustinvtran 🙇
